File: desktop/heartbeat_schedule.go
package main

import (
	"strconv"
	"strings"
	"time"
)

// parseInterval converts a string like "5m", "1h", "30s" to time.Duration.
// Suffix after '|' is stripped (e.g. "24h|daily@09:00" -> "24h").
// Empty or invalid strings return 0, nil (task will be skipped).
func parseInterval(s string) (time.Duration, error) {
	if idx := strings.Index(s, "|"); idx >= 0 {
		s = s[:idx]
	}
	if len(s) == 0 {
		return 0, nil
	}
	// Support common suffixed intervals
	switch s[len(s)-1] {
	case 's', 'm', 'h':
		return time.ParseDuration(s)
	default:
		// Try "Xm" as default assumption
		return time.ParseDuration(s + "m")
	}
}

// isCronExpr returns true when s looks like a valid 5-field cron expression
// (e.g. "0 * * * *", "*/15 * * * *", "0 9 * * 1-5"). Rejects expressions with
// out-of-range values (e.g. "99 * * * *") that would never match.
func isCronExpr(s string) bool {
	fields := strings.Fields(s)
	if len(fields) != 5 {
		return false
	}
	// min, hour, dom, month, dow — dom/month are 1-based (0 can never match
	// time.Day()/time.Month()), dow is 0-7 (7 = Sunday alias, matched in cronDue).
	limits := []int{59, 23, 31, 12, 7}
	mins := []int{0, 0, 1, 1, 0}
	for i, f := range fields {
		if f == "" {
			return false
		}
		for _, c := range f {
			if !strings.ContainsRune("0123456789*/-,", c) {
				return false
			}
		}
		// Reject out-of-range values in each field, plus zero/empty step
		// values ("*/0" never fires) and descending ranges ("5-1" never matches).
		for part := range strings.SplitSeq(f, ",") {
			part = strings.TrimSpace(part)
			base := part
			if stepBase, stepText, ok := strings.Cut(part, "/"); ok {
				step, err := strconv.Atoi(stepText)
				if err != nil || step < 1 {
					return false
				}
				base = stepBase
			}
			if base == "*" {
				continue
			}
			if loText, hiText, ok := strings.Cut(base, "-"); ok {
				lo, err1 := strconv.Atoi(loText)
				hi, err2 := strconv.Atoi(hiText)
				if err1 != nil || err2 != nil || lo < mins[i] || hi > limits[i] || lo > hi {
					return false
				}
			} else {
				v, err := strconv.Atoi(base)
				if err != nil || v < mins[i] || v > limits[i] {
					return false
				}
			}
		}
	}
	return true
}

// cronMatchField checks whether a single value matches a cron field pattern.
func cronMatchField(pattern string, value, minValue, maxValue int) bool {
	// Handle comma-separated lists
	for part := range strings.SplitSeq(pattern, ",") {
		part = strings.TrimSpace(part)
		if cronMatchSingle(part, value, minValue, maxValue) {
			return true
		}
	}
	return false
}

func cronMatchSingle(pattern string, value, minValue, maxValue int) bool {
	// Handle step values: */15, 1-10/2, 1/2. Wildcard steps are
	// anchored at the field minimum, which matters for 1-based fields.
	if rangePart, stepStr, ok := strings.Cut(pattern, "/"); ok {
		step, err := strconv.Atoi(stepStr)
		if err != nil || step <= 0 {
			return false
		}
		if rangePart == "*" {
			return value >= minValue && value <= maxValue && (value-minValue)%step == 0
		}
		// Range with step: 1-10/2
		if lowText, highText, ok := strings.Cut(rangePart, "-"); ok {
			low, _ := strconv.Atoi(lowText)
			high, _ := strconv.Atoi(highText)
			if value < low || value > high {
				return false
			}
			return (value-low)%step == 0
		}
		low, err := strconv.Atoi(rangePart)
		if err != nil || value < low || value > maxValue {
			return false
		}
		return (value-low)%step == 0
	}
	// Handle ranges: 1-5
	if lowText, highText, ok := strings.Cut(pattern, "-"); ok {
		low, err1 := strconv.Atoi(lowText)
		high, err2 := strconv.Atoi(highText)
		if err1 != nil || err2 != nil {
			return false
		}
		return value >= low && value <= high
	}
	// Handle wildcard
	if pattern == "*" {
		return true
	}
	// Handle literal value
	v, err := strconv.Atoi(pattern)
	if err != nil {
		return false
	}
	return v == value
}

// cronDue checks whether a 5-field cron expression should fire at the given time.
func cronDue(expr string, t time.Time) bool {
	if !isCronExpr(expr) {
		return false
	}
	fields := strings.Fields(expr)
	if len(fields) != 5 {
		return false
	}
	// Minute/hour/month must all match. When both day fields are restricted,
	// standard cron ORs them: "0 9 1 * 1" fires on the 1st or on Mondays,
	// rather than only when the 1st happens to be a Monday.
	domRestricted := fields[2] != "*"
	dowRestricted := fields[4] != "*"
	domMatch := cronMatchField(fields[2], t.Day(), 1, 31)
	// Standard cron allows 7 as a Sunday alias in the dow field (weekday is
	// 0-6 in Go); a pattern like "0 9 * * 7" must still fire on Sundays.
	weekday := int(t.Weekday())
	dowMatch := cronMatchField(fields[4], weekday, 0, 7) || (weekday == 0 && cronMatchField(fields[4], 7, 0, 7))
	dayMatch := !domRestricted || !dowRestricted
	if domRestricted && dowRestricted {
		dayMatch = domMatch || dowMatch
	} else if domRestricted {
		dayMatch = domMatch
	} else if dowRestricted {
		dayMatch = dowMatch
	}
	return cronMatchField(fields[0], t.Minute(), 0, 59) &&
		cronMatchField(fields[1], t.Hour(), 0, 23) &&
		dayMatch &&
		cronMatchField(fields[3], int(t.Month()), 1, 12)
}

func heartbeatTaskDueAt(t HeartbeatTask, now time.Time) bool {
	// Try cron expression first
	if isCronExpr(t.Interval) {
		if cronDue(t.Interval, now) {
			if t.LastRunAt == 0 || time.UnixMilli(t.LastRunAt).Before(now.Truncate(time.Minute)) {
				return true
			}
		}
		return false
	}

	if scheduled, ok := previousHeartbeatScheduleAt(t, now); ok {
		if t.CreatedAt != 0 && scheduled.Before(time.UnixMilli(t.CreatedAt)) {
			return false
		}
		if t.LastRunAt != 0 && !time.UnixMilli(t.LastRunAt).Before(scheduled) {
			return false
		}
		return !scheduled.After(now)
	}

	d, err := parseInterval(t.Interval)
	if err != nil || d <= 0 {
		return false
	}
	baseMillis := t.LastRunAt
	if baseMillis == 0 {
		baseMillis = t.CreatedAt
	}
	hasTimeWindow := t.TimeWindowStart != "" || t.TimeWindowEnd != ""
	if baseMillis == 0 {
		if hasTimeWindow {
			return heartbeatWithinTimeWindow(t, now)
		}
		return true
	}
	if now.Sub(time.UnixMilli(baseMillis)) < d {
		return false
	}

	// For interval-based tasks with a time window, check if current time
	// falls within the configured window. If outside, defer until the next
	// tick that falls within the window.
	if hasTimeWindow {
		return heartbeatWithinTimeWindow(t, now)
	}

	return true
}

// heartbeatWithinTimeWindow returns true when now falls within the task's
// configured time window. If the window is empty it returns true.
// Format: "HH:MM" in 24-hour clock; start inclusive, end exclusive.
func heartbeatWithinTimeWindow(t HeartbeatTask, now time.Time) bool {
	startH, startM, startOK := parseHeartbeatClock(t.TimeWindowStart)
	endH, endM, endOK := parseHeartbeatClock(t.TimeWindowEnd)

	if !startOK && !endOK {
		return true // no window configured
	}

	minutes := now.Hour()*60 + now.Minute()

	// If only start is set: allow from start to end of day
	if startOK && !endOK {
		return minutes >= startH*60+startM
	}

	// If only end is set: allow from midnight to end
	if !startOK && endOK {
		return minutes < endH*60+endM
	}

	startMin := startH*60 + startM
	endMin := endH*60 + endM

	if startMin < endMin {
		// Normal window: 09:00-17:00
		return minutes >= startMin && minutes < endMin
	}
	// Cross-midnight window: 22:00-06:00
	return minutes >= startMin || minutes < endMin
}

type heartbeatSchedule struct {
	kind     string
	days     []time.Weekday
	month    int
	day      int
	hour     int
	minute   int
	hasRules bool
}

func parseHeartbeatSchedule(interval string) (heartbeatSchedule, bool) {
	_, after, ok0 := strings.Cut(interval, "|")
	if !ok0 {
		return heartbeatSchedule{}, false
	}
	raw := strings.TrimSpace(after)
	if raw == "" {
		return heartbeatSchedule{}, false
	}
	at := "09:00"
	if parts := strings.SplitN(raw, "@", 2); len(parts) == 2 {
		raw = parts[0]
		at = parts[1]
	}
	hour, minute, ok := parseHeartbeatClock(at)
	if !ok {
		return heartbeatSchedule{}, false
	}
	kind := raw
	rule := ""
	if parts := strings.SplitN(raw, ":", 2); len(parts) == 2 {
		kind = parts[0]
		rule = parts[1]
	}
	s := heartbeatSchedule{kind: kind, hour: hour, minute: minute, hasRules: true}
	switch kind {
	case "daily":
		return s, true
	case "weekly", "biweekly":
		for part := range strings.SplitSeq(rule, ",") {
			if wd, ok := parseHeartbeatWeekday(part); ok {
				s.days = append(s.days, wd)
			}
		}
		return s, len(s.days) > 0
	case "monthly":
		s.day = parsePositiveInt(rule, 1)
		return s, true
	case "yearly":
		parts := strings.SplitN(rule, "-", 2)
		s.month = parsePositiveInt(firstString(parts), 1)
		s.day = 1
		if len(parts) == 2 {
			s.day = parsePositiveInt(parts[1], 1)
		}
		if s.month < 1 {
			s.month = 1
		}
		if s.month > 12 {
			s.month = 12
		}
		return s, true
	default:
		return heartbeatSchedule{}, false
	}
}

func previousHeartbeatScheduleAt(t HeartbeatTask, now time.Time) (time.Time, bool) {
	s, ok := parseHeartbeatSchedule(t.Interval)
	if !ok || !s.hasRules {
		return time.Time{}, false
	}
	switch s.kind {
	case "daily":
		candidate := dateAt(now.Year(), now.Month(), now.Day(), s.hour, s.minute, now.Location())
		if candidate.After(now) {
			candidate = candidate.AddDate(0, 0, -1)
		}
		return candidate, true
	case "weekly":
		return previousHeartbeatWeeklyAt(s, now, 7, time.Time{})
	case "biweekly":
		anchor := heartbeatScheduleAnchor(t, now)
		return previousHeartbeatWeeklyAt(s, now, 14, anchor)
	case "monthly":
		return previousHeartbeatMonthlyAt(s, now), true
	case "yearly":
		return previousHeartbeatYearlyAt(s, now), true
	default:
		return time.Time{}, false
	}
}

func previousHeartbeatWeeklyAt(s heartbeatSchedule, now time.Time, windowDays int, anchor time.Time) (time.Time, bool) {
	var best time.Time
	for offset := range windowDays {
		day := now.AddDate(0, 0, -offset)
		for _, wd := range s.days {
			if day.Weekday() != wd {
				continue
			}
			candidate := dateAt(day.Year(), day.Month(), day.Day(), s.hour, s.minute, now.Location())
			if candidate.After(now) {
				continue
			}
			if !anchor.IsZero() && weeksBetween(weekStart(anchor), weekStart(candidate))%2 != 0 {
				continue
			}
			if best.IsZero() || candidate.After(best) {
				best = candidate
			}
		}
	}
	return best, !best.IsZero()
}

func previousHeartbeatMonthlyAt(s heartbeatSchedule, now time.Time) time.Time {
	candidate := monthlyCandidate(now.Year(), now.Month(), s.day, s.hour, s.minute, now.Location())
	if candidate.After(now) {
		prev := now.AddDate(0, -1, 0)
		candidate = monthlyCandidate(prev.Year(), prev.Month(), s.day, s.hour, s.minute, now.Location())
	}
	return candidate
}

func previousHeartbeatYearlyAt(s heartbeatSchedule, now time.Time) time.Time {
	month := time.Month(s.month)
	candidate := monthlyCandidate(now.Year(), month, s.day, s.hour, s.minute, now.Location())
	if candidate.After(now) {
		candidate = monthlyCandidate(now.Year()-1, month, s.day, s.hour, s.minute, now.Location())
	}
	return candidate
}

func heartbeatScheduleAnchor(t HeartbeatTask, now time.Time) time.Time {
	if t.CreatedAt != 0 {
		return time.UnixMilli(t.CreatedAt)
	}
	if t.LastRunAt != 0 {
		return time.UnixMilli(t.LastRunAt)
	}
	return now
}

func parseHeartbeatClock(s string) (int, int, bool) {
	parts := strings.SplitN(strings.TrimSpace(s), ":", 2)
	if len(parts) != 2 {
		return 0, 0, false
	}
	hour := parsePositiveInt(parts[0], -1)
	minute := parsePositiveInt(parts[1], -1)
	if hour < 0 || hour > 23 || minute < 0 || minute > 59 {
		return 0, 0, false
	}
	return hour, minute, true
}

func parseHeartbeatWeekday(s string) (time.Weekday, bool) {
	switch strings.ToLower(strings.TrimSpace(s)) {
	case "sun":
		return time.Sunday, true
	case "mon":
		return time.Monday, true
	case "tue":
		return time.Tuesday, true
	case "wed":
		return time.Wednesday, true
	case "thu":
		return time.Thursday, true
	case "fri":
		return time.Friday, true
	case "sat":
		return time.Saturday, true
	default:
		return time.Sunday, false
	}
}

func parsePositiveInt(s string, fallback int) int {
	s = strings.TrimSpace(s)
	if s == "" {
		return fallback
	}
	n := 0
	for _, r := range s {
		if r < '0' || r > '9' {
			return fallback
		}
		n = n*10 + int(r-'0')
	}
	return n
}

func firstString(values []string) string {
	if len(values) == 0 {
		return ""
	}
	return values[0]
}

func dateAt(year int, month time.Month, day, hour, minute int, loc *time.Location) time.Time {
	return time.Date(year, month, day, hour, minute, 0, 0, loc)
}

func monthlyCandidate(year int, month time.Month, day, hour, minute int, loc *time.Location) time.Time {
	if day < 1 {
		day = 1
	}
	if max := daysInMonth(year, month, loc); day > max {
		day = max
	}
	return dateAt(year, month, day, hour, minute, loc)
}

func daysInMonth(year int, month time.Month, loc *time.Location) int {
	return time.Date(year, month+1, 0, 0, 0, 0, 0, loc).Day()
}

func weekStart(t time.Time) time.Time {
	dayOffset := (int(t.Weekday()) + 6) % 7
	base := dateAt(t.Year(), t.Month(), t.Day(), 0, 0, t.Location())
	return base.AddDate(0, 0, -dayOffset)
}

func weeksBetween(a, b time.Time) int {
	// Compare civil dates in UTC rather than elapsed local hours. Local Monday
	// midnights spanning DST can be 167 or 169 hours apart despite being exactly
	// one calendar week apart.
	aDay := time.Date(a.Year(), a.Month(), a.Day(), 0, 0, 0, 0, time.UTC)
	bDay := time.Date(b.Year(), b.Month(), b.Day(), 0, 0, 0, 0, time.UTC)
	if bDay.Before(aDay) {
		aDay, bDay = bDay, aDay
	}
	return int(bDay.Sub(aDay) / (7 * 24 * time.Hour))
}

File: desktop/frontend/src/custom/features/heartbeat/heartbeat.schedule.ts
import type { HeartbeatTask } from "./heartbeat.types";

const INTERVAL_MS: Record<"s" | "m" | "h", number> = {
  s: 1000,
  m: 60_000,
  h: 3_600_000,
};

const WEEKDAYS: Record<string, number> = {
  sun: 0,
  mon: 1,
  tue: 2,
  wed: 3,
  thu: 4,
  fri: 5,
  sat: 6,
};

interface CalendarSchedule {
  kind: "daily" | "weekly" | "biweekly" | "monthly" | "yearly";
  days: number[];
  month: number;
  day: number;
  hour: number;
  minute: number;
}

function heartbeatIntervalMs(interval?: string): number | null {
  const clean = (interval || "").replace(/\|.*$/, "");
  const match = clean.match(/^(\d+)([smh])$/);
  if (!match) return null;
  return parseInt(match[1], 10) * INTERVAL_MS[match[2] as "s" | "m" | "h"];
}

function heartbeatClockMinutes(value?: string): number | null {
  const match = (value || "").match(/^(\d{2}):(\d{2})$/);
  if (!match) return null;
  const hour = parseInt(match[1], 10);
  const minute = parseInt(match[2], 10);
  if (hour < 0 || hour > 23 || minute < 0 || minute > 59) return null;
  return hour * 60 + minute;
}

function dateAtMinutes(base: Date, minutes: number): Date {
  const date = new Date(base);
  date.setHours(Math.floor(minutes / 60), minutes % 60, 0, 0);
  return date;
}

function heartbeatWithinWindow(date: Date, start: number | null, end: number | null): boolean {
  if (start === null && end === null) return true;
  const minutes = date.getHours() * 60 + date.getMinutes();
  if (start !== null && end === null) return minutes >= start;
  if (start === null && end !== null) return minutes < end;
  if (start === end) return true;
  if (start! < end!) return minutes >= start! && minutes < end!;
  return minutes >= start! || minutes < end!;
}

function nextHeartbeatWindowTime(from: Date, start: number | null, end: number | null): Date {
  if (heartbeatWithinWindow(from, start, end)) return from;
  if (start !== null && end === null) return dateAtMinutes(from, start);
  if (start === null && end !== null) {
    const next = new Date(from);
    next.setDate(next.getDate() + 1);
    next.setHours(0, 0, 0, 0);
    return next;
  }
  const minutes = from.getHours() * 60 + from.getMinutes();
  if (start! < end! && minutes < start!) return dateAtMinutes(from, start!);
  if (start! > end! && minutes < start! && minutes >= end!) return dateAtMinutes(from, start!);
  const next = dateAtMinutes(from, start!);
  next.setDate(next.getDate() + 1);
  return next;
}

function positiveInt(value: string, fallback: number): number {
  return /^\d+$/.test(value.trim()) ? parseInt(value, 10) : fallback;
}

function parseCalendarScheduleImpl(interval?: string): CalendarSchedule | null {
  const separator = (interval || "").indexOf("|");
  if (separator < 0) return null;
  const suffix = interval!.slice(separator + 1).trim();
  const clockSeparator = suffix.indexOf("@");
  const rulePart = clockSeparator < 0 ? suffix : suffix.slice(0, clockSeparator);
  const clock = clockSeparator < 0 ? "09:00" : suffix.slice(clockSeparator + 1);
  const clockMinutes = heartbeatClockMinutes(clock);
  if (clockMinutes === null) return null;
  const ruleSeparator = rulePart.indexOf(":");
  const kind = ruleSeparator < 0 ? rulePart : rulePart.slice(0, ruleSeparator);
  const rule = ruleSeparator < 0 ? "" : rulePart.slice(ruleSeparator + 1);
  if (!(["daily", "weekly", "biweekly", "monthly", "yearly"] as const).includes(kind as CalendarSchedule["kind"])) {
    return null;
  }
  const schedule: CalendarSchedule = {
    kind: kind as CalendarSchedule["kind"],
    days: [],
    month: 1,
    day: 1,
    hour: Math.floor(clockMinutes / 60),
    minute: clockMinutes % 60,
  };
  if (schedule.kind === "weekly" || schedule.kind === "biweekly") {
    schedule.days = rule.split(",").map((day) => WEEKDAYS[day.trim().toLowerCase()]).filter((day) => day !== undefined);
    if (schedule.days.length === 0) return null;
  } else if (schedule.kind === "monthly") {
    schedule.day = positiveInt(rule, 1);
  } else if (schedule.kind === "yearly") {
    const [month, day] = rule.split("-", 2);
    schedule.month = Math.min(12, Math.max(1, positiveInt(month, 1)));
    schedule.day = positiveInt(day || "", 1);
  }
  return schedule;
}

function calendarDate(year: number, month: number, day: number, schedule: CalendarSchedule): Date {
  const maxDay = new Date(year, month + 1, 0).getDate();
  return new Date(year, month, Math.min(Math.max(day, 1), maxDay), schedule.hour, schedule.minute, 0, 0);
}

function weekStart(date: Date): Date {
  const start = new Date(date.getFullYear(), date.getMonth(), date.getDate());
  start.setDate(start.getDate() - ((start.getDay() + 6) % 7));
  return start;
}

function civilDayNumber(date: Date): number {
  return Math.floor(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()) / 86_400_000);
}

// Exported for the panel's cycle interval math (nextCycleRunAt) so both frontend
// paths share one calendar implementation and cannot drift from each other.
export function parseCalendarSchedule(interval?: string): CalendarSchedule | null {
  return parseCalendarScheduleImpl(interval);
}

export function calendarDateFor(year: number, month: number, day: number, schedule: CalendarSchedule): Date {
  return calendarDate(year, month, day, schedule);
}

export function nextCalendarRunImpl(schedule: CalendarSchedule, after: Date, anchor: Date): Date {
  return nextCalendarRun(schedule, after, anchor);
}

function nextCalendarRun(schedule: CalendarSchedule, after: Date, anchor: Date): Date {
  if (schedule.kind === "daily") {
    const candidate = calendarDate(after.getFullYear(), after.getMonth(), after.getDate(), schedule);
    if (candidate.getTime() <= after.getTime()) candidate.setDate(candidate.getDate() + 1);
    return candidate;
  }
  if (schedule.kind === "weekly" || schedule.kind === "biweekly") {
    const searchDays = schedule.kind === "biweekly" ? 21 : 7;
    for (let offset = 0; offset <= searchDays; offset += 1) {
      const day = new Date(after.getFullYear(), after.getMonth(), after.getDate());
      day.setDate(day.getDate() + offset);
      if (!schedule.days.includes(day.getDay())) continue;
      const candidate = calendarDate(day.getFullYear(), day.getMonth(), day.getDate(), schedule);
      if (candidate.getTime() <= after.getTime()) continue;
      if (schedule.kind === "biweekly") {
        const weeks = Math.floor(Math.abs(civilDayNumber(weekStart(candidate)) - civilDayNumber(weekStart(anchor))) / 7);
        if (weeks % 2 !== 0) continue;
      }
      return candidate;
    }
  }
  if (schedule.kind === "monthly") {
    let candidate = calendarDate(after.getFullYear(), after.getMonth(), schedule.day, schedule);
    if (candidate.getTime() <= after.getTime()) {
      const nextMonth = new Date(after.getFullYear(), after.getMonth() + 1, 1);
      candidate = calendarDate(nextMonth.getFullYear(), nextMonth.getMonth(), schedule.day, schedule);
    }
    return candidate;
  }
  if (schedule.kind === "yearly") {
    let candidate = calendarDate(after.getFullYear(), schedule.month - 1, schedule.day, schedule);
    if (candidate.getTime() <= after.getTime()) candidate = calendarDate(after.getFullYear() + 1, schedule.month - 1, schedule.day, schedule);
    return candidate;
  }
  return after;
}

export function heartbeatNextRunAt(
  task: Pick<HeartbeatTask, "interval" | "lastRunAt" | "createdAt" | "timeWindowStart" | "timeWindowEnd">,
  now = Date.now(),
): number | null {
  if (!task.lastRunAt) return null;
  const intervalMs = heartbeatIntervalMs(task.interval);
  if (intervalMs === null) return null;
  const schedule = parseCalendarSchedule(task.interval);
  if (schedule) {
    const after = new Date(task.lastRunAt);
    const anchor = new Date(task.createdAt || task.lastRunAt);
    return nextCalendarRun(schedule, after, anchor).getTime();
  }
  const rawNext = task.lastRunAt + intervalMs;
  const start = heartbeatClockMinutes(task.timeWindowStart);
  const end = heartbeatClockMinutes(task.timeWindowEnd);
  if (start === null && end === null) return rawNext;
  const candidate = new Date(Math.max(rawNext, now));
  return nextHeartbeatWindowTime(candidate, start, end).getTime();
}
